README.md
# energyusage

A Python package that measures the environmental impact of computation. Provides a function to
evaluate the energy usage and related carbon emissions of another function.
Emissions are calculated based on the user's location via the GeoJS API and that location's
energy mix data (sources: US E.I.A and eGRID for the year 2016).

## Installation

To install, simply `$ pip install energyusage`.

## Usage

To evaluate the emissions of a function, just call `energyusage.evaluate` with the function
name and the arguments it requires.

```python
import energyusage

# user function to be evaluated
def recursive_fib(n):
    if (n <= 2): return 1
    else: return recursive_fib(n-1) + recursive_fib(n-2)

energyusage.evaluate(recursive_fib, 40, pdf=True)
# returns 102,334,155
```
It will return the value of your function, while also printing out the energy usage report on the command line.
Optional keyword arguments:
* `pdf`(default = `False`): generates a PDF report, alongside the command-line utility
* `powerLoss` (default = `0.8`): accounts for PSU loss, can be set by user if known for higher accuracy of results
* `energyOutput` (default = `False`): prints amount of energy used by the process and time taken. The order is time, energy used, return of function
* `printToScreen` (default = `True`): controls whether there is a terminal printout of the package running
* `energyOutput` (default = `False`): determines whether the energy used and time taken are output. When set to true the order is `time used`, `energy used`, `return value of function`.
* `locations` (default = `["Mongolia", "Iceland", "Switzerland"]`): allows selecting the countries in the emissions comparison section for the terminal printout and pdf. These can be set to the name of any country or US state.
* `year` (default = `2016`): controls the year for the data. Default is `2016` as that is currently the most recent year of data from both of our sources. Note that only this year of data is included in the package installation but more can be added in a process described later.

### Energy Report
The report that will be printed out will look like the one below. The second and third lines will show a real-time reading that disappears once the process has finished evaluating.
```
Location:                                                           Pennsylvania
--------------------------------------------------------------------------------
-------------------------------  Final Readings  -------------------------------
--------------------------------------------------------------------------------
Average baseline wattage:                                             1.86 watts
Average total wattage:                                               19.42 watts
Average process wattage:                                             17.56 watts
Process duration:                                                        0:00:01
--------------------------------------------------------------------------------
-------------------------------   Energy Data    -------------------------------
--------------------------------------------------------------------------------
                           Energy mix in Pennsylvania                           
Coal:                                                                     25.42%
Oil:                                                                       0.17%
Natural Gas:                                                              31.64%
Low Carbon:                                                               42.52%
--------------------------------------------------------------------------------
-------------------------------    Emissions     -------------------------------
--------------------------------------------------------------------------------
Effective emission:                                              4.05e-06 kg CO2
Equivalent miles driven:                                          1.66e-12 miles
Equivalent minutes of 32-inch LCD TV watched:                   2.51e-03 minutes
Percentage of CO2 used in a US household/day:                          1.33e-12%
--------------------------------------------------------------------------------
------------------------- Assumed Carbon Equivalencies -------------------------
--------------------------------------------------------------------------------
Coal:                                                      995.725971 kg CO2/MWh
Petroleum:                                                816.6885263 kg CO2/MWh
Natural gas:                                              743.8415916 kg CO2/MWh
Low carbon:                                                         0 kg CO2/MWh
--------------------------------------------------------------------------------
-------------------------     Emissions Comparison     -------------------------
--------------------------------------------------------------------------------
                      Quantities below expressed in kg CO2                      
        US                      Europe                  Global minus US/Europe
Max:    Wyoming        9.59e-06 Kosovo         9.85e-06 Mongolia        9.64e-06
Median: Tennessee      4.70e-06 Ukraine        6.88e-06 Korea, South    7.87e-06
Min:    Vermont        2.69e-07 Iceland        1.77e-06 Bhutan          1.10e-06
--------------------------------------------------------------------------------
--------------------------------------------------------------------------------
Process used:                                                       1.04e-05 kWh
```
The report is divided into several sections.
* **Final Readings**: Presents an average of:
	* *Average baseline wattage*: your computer's average power usage minus the process, ran for 10 seconds before starting your process
	* *Average total wattage*: your computer's average power usage while the process runs
	* *Average process usage*: the difference between the baseline and total, highlighting the usage solely from the specific process you evaluated
	* *Process duration*: how long your program ran for

* **Energy Data**: The energy mix of the location.

* **Emissions**: The effective CO<sub>2</sub> emissions of running the program one time and some real-world equivalents to those emissions.

* **Assumed Carbon Equivalencies**: The formulas used to convert from kWh to CO<sub>2</sub> based on the energy mix of the location (for international locations, see below for more information).

* **Emissions Comparison**: What the emissions would be for the same energy used in a representative group of US states and countries. Note that if these locations are specified as described below these default values are not shown. 

* **Process used**: The amount of energy running the program used in total.

The PDF report contains the same sections, but does not include the process duration or the emissions comparison momentarily.

## Methodology
### Power Measurement
#### CPU
We calculate CPU power usage via the RAPL (Running Average Power Limit) interfaces found on Intel processors. These are non-architectural model-specific registers that provide power-related information
about the CPU. They are used primarily for limiting power consumption, but the Energy Status
register (MSR_PKG_ENERGY_STATUS) allows for power measurement.

The RAPL interface differentiates between several domains, based on the number of processors. For a single processor machine:
  * Package
  * Power planes:
      * Core
      * Uncore
  * DRAM

For a machine with multiple processors:
  * Package 0
  * Package 1
  * ...
  * Package n
  * DRAM

Presently, we use the Package domain (or a sum of all of the domains, for a multi-processor machine), which represents the complete processor package.

As outlined by [Vince Weaver](http://web.eece.maine.edu/~vweaver/projects/rapl/), there are multiple ways to access the RAPL interface data, namely:
 * Using perf_event interface
 * Reading the underlying MSR
 * Reading the files under `/sys/class/powercap/intel-rapl/`

We elected to use the final method because it is the only one that does not require sudo access.  We read the `energy_uj.txt` files inside the package folder(s) `intel-rapl:*`. These files represent the energy used in microjoules, and they update roughly every millisecond. The value in the file increases to the point of overflow and then resets. We take 2 readings with a delay in-between, and then calculate the wattage based on the difference (energy) and the delay (time). To avoid errors due to the reset of the file, we discard negative values.

For more information on the RAPL interface, consult the [Intel® 64 and IA-32 Architectures Software Developer's Manual](https://software.intel.com/sites/default/files/managed/39/c5/325462-sdm-vol-1-2abcd-3abcd.pdf).

#### GPU
To the package measurement we also add the power usage of the GPU for machines that have an Nvidia GPU that support the NVIDIA-smi program.

The NVIDIA-smi is a command-line utility that allows for the querying of information about the GPU. If the GPU is identified as valid, we use the built-in method to query the current wattage, and then convert the output into a float.

More  information on NVIDIA-smi can be found on the [Nvidia website](https://developer.nvidia.com/nvidia-system-management-interface).


### Calculating CO<sub>2</sub> Emissions
#### Location
In order to accurately calculate the CO₂ emissions associated with the computational power used, we determine the geographical location of the user via their IP address with the help of the [GeoJS](https://www.geojs.io/) API.  If the location cannot be determined, we use the United States as the default.

Location is especially important as the emissions differ based on the country's (and, in the case of the United States, the state's) energy mix.

#### Energy Mix Information
We obtained international energy mix data from the [U.S. Energy Information Administration data](https://www.eia.gov/beta/international/data/browser/#/?pa=0000000010000000000000000000000000000000000000000000000000u&c=ruvvvvvfvsujvv1vrvvvvfvvvvvvfvvvou20evvvfvrvvvvvvurs&ct=0&vs=INTL.44-2-AFG-QBTU.A&cy=2016&vo=0&v=H&start=2014&end=2016) for the year 2016. Specifically, we looked at the energy consumption of countries worldwide, broken down by energy source. For the data points labeled *(s)* (meaning that the value is too small for the number of decimal places shown), we approximated those amounts to 0. No data was available for, and thus we removed from consideration, the following: Former Czechoslovakia, Former Serbia and Montenegro, Former U.S.S.R., Former Yugoslavia, Hawaiian Trade Zone, East Germany and West Germany.

Our United States energy mix and emissions data was obtained from the [U.S. Environmental Protection Agency eGRID data](https://www.epa.gov/sites/production/files/2018-02/egrid2016_summarytables.xlsx) for the year 2016. We used the *State Resource Mix* section for displaying the energy mix, and the *State Output Emission Rates* section for calculating emissions in the United States. We did not use the *otherFossil* data as the values were predominantly 0 (and in cases in which the value was nonzero, it was below 1%).

As of July 2019, the most recent eGRID data was from the year 2016. We elected to use 2016 U.S. E.I.A. data for consistency between the data sources.

#### Conversion to CO<sub>2</sub>
Since the international data only contained an energy mix, and no emission data, we reverse-engineered the formulas used in the eGRID data. This gives us additionally consistency between the separate datasets.
* *Coal*: 2195.20 lbs CO<sub>2</sub>/MWh = 995.725971 kg CO<sub>2</sub>/MWh
* *Petroleum*: lbs CO<sub>2</sub>/MWh = 816.6885263 kg CO<sub>2</sub>/MWh
* *Natural gas*: 1639.89 lbs CO<sub>2</sub>/MWh = 743.8415916 kg CO<sub>2</sub>/MWh

#### Using Different Years of Data
In case one wishes to compare energy usage between different years of data, we have included a script to allow for adding other years. If you navigate to the package directory and go into the `data` folder, you can use `raw_to_json.py`. First, you need to download the US and international data in the years of your choice from the links above and place them in `data/raw/"year of the data"` after creating the required year folder. Then, run the script with a flag for that year (for example, `python raw_to_json.py -2016`). This will allow selecting that year when using package in the future by using the `year` optional argument for `evaluate`.

## Related Work
* In their paper [*Energy and Policy Considerations for Deep Learning in NLP*](https://arxiv.org/abs/1906.02243), Strubell et. al not only analyze the computational power needed for training deep learning models in NLP, but further convert the data into carbon emissions and cost. Our tool aims to facilitate this analysis for developers in a single package. We do not consider cost, instead choosing to focus solely on the environmental impact. Further, we do not focus on a specific computational area. We also extend their analysis of carbon emissions by including international data on energy consumption and CO<sub>2</sub> emissions for localized analysis of the carbon footprint of the tested program.

## Limitations
* Due to the methods in which the energy measurement is being done (through the Intel RAPL
interface and NVIDIA-smi), our package is only available on Linux kernels that have the
RAPL interface and/or machines with an Nvidia GPU.

*  A country’s overall energy consumption mix is not necessarily representative of the mix of energy sources used to produce electricity (and even electricity production is not necessarily representative of electricity consumption due to imports/exports). However, the E.I.A. data is the most geographically comprehensive that we have found. We are working on obtaining even more accurate data.


## Acknowledgements
We would like to thank [Jon Wilson](https://www.haverford.edu/users/jwilson) for all his valuable insight with regards to the environmental aspect of our project.


changelog.md
# changelog

v0.0.2
* removed emissions from return of `evaluate`, now just returns value of user function
* added `Average total wattage` to report to distinguish between baseline+process and just process
* added to README to make it comprehensive
* added new carbon equivalent: minutes of TV watched

v0.0.3
* minor bug fixes

v0.0.4
* updated conversion formulas
* added powerLoss parameter

v0.0.5
* added Emissions Comparison section to command-line report
* minor bug fixes

v.0.0.6
* fixed multiprocessing bug that would cause deadlock

v.0.0.7
* added option of getting emissions value from `evaluate` via energyOutput parameter

v.0.0.8 - 0.0.12
* Added methods for accessing other years of data
* Fleshed out the pdf report
* Added several new flags for the evaluate function
* General bug fixes

v.0.0.13
* Added default graphs to pdf when locations optional parameter is not set


energyusage/utils.py
import math
import sys
import operator
import time
import json
import os
import re
import statistics
import subprocess

import energyusage.convert as convert
import energyusage.locate as locate
from energyusage.RAPLFile import RAPLFile

printToScreenGlobal = True
BASE = "/sys/class/powercap/"
DELAY = .1 # in seconds
DIR_PATH = os.path.dirname(os.path.realpath(__file__))

""" MEASUREMENT UTILS """

def read(file):
    """ Opens file and reads energy measurement """
    if file == "":
        return 0
    else:
        with open(file, 'r') as f:
            return convert.to_joules(int(f.read()))

def average_files(raplfiles):
    for file in raplfiles:
        file.process_average = statistics.mean(file.process)
        file.baseline_average = statistics.mean(file.baseline)
    return raplfiles

def measure(file, delay=1):
    """ Measures the energy output of FILE """

    start, end = 0, 0
    start = read(file)
    time.sleep(delay)
    end = read(file)

    return end-start

def get_process_average(raplfiles, multiple_cpus, gpu):
    total = 0
    if multiple_cpus:

        for file in raplfiles:
            if "CPU" in file.name:
                total+= file.process_average
    else:
        for file in raplfiles:
            if file.name == "Package":
                total+=file.process_average
    return total + gpu

def get_baseline_average(raplfiles, multiple_cpus, gpu):
    total = 0
    if multiple_cpus:
        for file in raplfiles:
            if "CPU" in file.name:
                total+= file.baseline_average
    else:
        for file in raplfiles:
            if file.name == "Package":
                total+=file.baseline_average
    return total + gpu

def get_total(raplfiles, multiple_cpus):
    total = 0
    if multiple_cpus:
        for file in raplfiles:
            if "CPU" in file.name:
                total+= file.recent
    else:
        for file in raplfiles:
            if file.name == "Package":
                total =  file.recent
    if (total):
        return total
    return 0

def update_files(raplfiles, process = False):
    if process:
        for file in raplfiles:
            if file.recent >= 0:
                file.process.append(file.recent)
    else:
        for file in raplfiles:
            if file.recent >= 0:
                file.baseline.append(file.recent)
    return raplfiles

def start(raplfile):
    measurement = read(raplfile.path)
    raplfile.recent = measurement
    return raplfile

def end(raplfile, delay):
    measurement = read(raplfile.path)
    raplfile.recent = (measurement -raplfile.recent) / delay
    return raplfile

def measure_files(files, delay = 1):
    """ Measures the energy output of all packages which should give total power usage

    Parameters:
        files (list): list of RAPLFiles
        delay (int): RAPL file reading rate in ms

    Returns:
        files (list): list of RAPLfiles with updated measurements
    """

    files = list(map(start, files))
    time.sleep(delay)
    files = list(map(lambda x: end(x, delay), files)) # need lambda to pass in delay
    return files

def reformat(name, multiple_cpus):
    """ Renames the RAPL files for better readability/understanding """
    if 'package' in name:
        if multiple_cpus:
            name = "CPU" + name[-1] # renaming it to CPU-x
        else:
            name = "Package"
    if name == 'core':
        name = "CPU"
    elif name == 'uncore':
        name = "GPU"
    elif name == 'dram':
        name = name.upper()
    return name

def get_files():
    """ Gets all the RAPL files with their names on the machine

        Returns:
            filenames (list): list of RAPLFiles
    """
    # Removing the intel-rapl folder that has no info
    files = list(filter(lambda x: ':' in x, os.listdir(BASE)))
    names = {}
    cpu_count = 0
    multiple_cpus = False
    for file in files:
        if (re.fullmatch("intel-rapl:.", file)):
            cpu_count += 1

    if cpu_count > 1:
        multiple_cpus = True

    for file in files:
        path = BASE + '/' + file + '/name'
        with open(path) as f:
           name = f.read()[:-1]
           renamed = reformat(name, multiple_cpus)
        names[renamed] = BASE + file + '/energy_uj'

    filenames = []
    for name, path in names.items():
        name = RAPLFile(name, path)
        filenames.append(name)

    return filenames, multiple_cpus


# from realpython.com/python-rounding
def round_up(n, decimals=4):
    """ Rounds up if digit is >= 5 """

    multiplier = 10 ** decimals
    return math.floor(n*multiplier + 0.5) / multiplier


""" LOGGING UTILS """

def log_header(text):
    if len(text) > 16:
        sys.stdout.write("-"*80 + "\n" + "-"*25 + " {:^28} ".format(text) +
            "-"*25 + "\n" + "-"*80+ "\n")
    else:
        sys.stdout.write("-"*80 + "\n" + "-"*31 + " {:^16} ".format(text) +
            "-"*31 + "\n" + "-"*80+ "\n")

# from https://stackoverflow.com/a/52590238
def delete_last_lines():
    # Moves cursor up one line
    sys.stdout.write('\x1b[1A')
    sys.stdout.write('\x1b[1A')

def newline():
    sys.stdout.write('\n')

def setGlobal(printToScreen):
    global printToScreenGlobal
    printToScreenGlobal = printToScreen

def log(*args):
    if (re.search("Package|CPU.*|GPU|DRAM", args[0])):
        measurement = args[1]
        sys.stdout.write("\r{:<24} {:>49.2f} {:5<}".format(args[0]+":", measurement, "watts"))

    if args[0] == "Baseline wattage":
        measurement = args[1]
        sys.stdout.write("\r{:<24} {:>49.2f} {:5<}".format(args[0]+":", measurement, "watts"))

    elif args[0] == "Process wattage":
        measurement = args[1]
        sys.stdout.write("\r{:<17} {:>56.2f} {:5<}".format(args[0]+":", measurement, "watts"))

    elif args[0] == "Final Readings":
        newline()
        baseline_average, process_average, difference_average, timedelta = args[1], args[2], args[3], args[4]
        delete_last_lines()
        log_header(args[0])
        sys.stdout.write("{:<25} {:>48.2f} {:5<}\n".format("Average baseline wattage:", baseline_average, "watts"))
        sys.stdout.write("{:<25} {:>48.2f} {:5<}\n".format("Average total wattage:", process_average, "watts"))
        sys.stdout.write("{:<25} {:>48.2f} {:5<}\n".format("Average process wattage:", difference_average, "watts"))
        sys.stdout.write("{:<17} {:>62}\n".format("Process duration:", timedelta))

    elif args[0] == "Energy Data":
        location = args[2]
        log_header('Energy Data')
        if location == "Unknown" or locate.in_US(location):
            coal, oil, gas, low_carbon = args[1]
            if location == "Unknown":
                location = "United States"
                sys.stdout.write("{:^80}\n{:<13}{:>66.2f}%\n{:<13}{:>66.2f}%\n{:<13}{:>66.2f}%\n"
                    "{:<13}{:>66.2f}%\n".format("Location unknown, default energy mix in "+location+":", "Coal:", coal, "Oil:", oil,
                    "Natural Gas:", gas, "Low Carbon:", low_carbon))
            elif locate.in_US(location):
                sys.stdout.write("{:^80}\n{:<13}{:>66.2f}%\n{:<13}{:>66.2f}%\n{:<13}{:>66.2f}%\n"
                    "{:<13}{:>66.2f}%\n".format("Energy mix in "+location, "Coal:", coal, "Oil:", oil,
                    "Natural Gas:", gas, "Low Carbon:", low_carbon))
        else:
            coal, natural_gas, petroleum, low_carbon = args[1]
            sys.stdout.write("{:^80}\n{:<13}{:>66.2f}%\n{:<13}{:>66.2f}%\n{:<13}{:>66.2f}%\n"
                    "{:<13}{:>66.2f}%\n".format("Energy mix in "+location, "Coal:", coal, "Petroleum:", petroleum,
                    "Natural Gas:", natural_gas, "Low Carbon:", low_carbon))

    elif args[0] == "Emissions":
        emission = args[1]
        log_header('Emissions')
        sys.stdout.write("{:<19}{:>54.2e} kg CO2\n".format("Effective emission:", \
            emission))
        sys.stdout.write("{:<24}{:>50.2e} miles\n".format("Equivalent miles driven:", \
            convert.carbon_to_miles(emission)))
        sys.stdout.write("{:<45}{:>27.2e} minutes\n".format("Equivalent minutes of 32-inch LCD TV watched:", \
            convert.carbon_to_tv(emission)))
        sys.stdout.write("{:<45}{:>34.2e}%\n".format("Percentage of CO2 used in a US"
        " household/day:",convert.carbon_to_home(emission)))

    elif args[0] == "Assumed Carbon Equivalencies":
        log_header('Assumed Carbon Equivalencies')
        sys.stdout.write("{:<14} {:>65}\n".format("Coal:", "995.725971 kg CO2/MWh"))
        sys.stdout.write("{:<14} {:>65}\n".format("Petroleum:", "816.6885263 kg CO2/MWh"))
        sys.stdout.write("{:<14} {:>65}\n".format("Natural gas:", "743.8415916 kg CO2/MWh"))
        sys.stdout.write("{:<14} {:>65}\n".format("Low carbon:", "0 kg CO2/MWh"))
    elif args[0] == "Emissions Comparison":
        log_header('Emissions Comparison')
        emissions = args[1]
        for location, emission in emissions:
            sys.stdout.write("{:<19}{:>54.2e} kg CO2\n".format(location+":", emission))

     #OLD VERSION: US, EU, Rest comparison
    elif args[0] == "Emissions Comparison default":
         log_header('Emissions Comparison')
         max_global, median_global, min_global, max_europe, median_europe, min_europe, \
         max_us, median_us, min_us = args[1:]
         sys.stdout.write("{:^80}\n".format("Quantities below expressed in kg CO2"))
         sys.stdout.write("{:8}{:<23} {:<23} {:<22}\n".format("", "US", "Europe", \
            "Global minus US/Europe"))
         sys.stdout.write("{:<7} {:<13}{:>10.2e} {:<13}{:>10.2e} {:<14}{:>10.2e}\n".format("Max:", max_us[0], max_us[1], \
             max_europe[0], max_europe[1], max_global[0], max_global[1]))
         sys.stdout.write("{:<7} {:<13}{:>10.2e} {:<13}{:>10.2e} {:<14}{:>10.2e}\n".format("Median:", median_us[0], median_us[1], \
             median_europe[0], median_europe[1], median_global[0], median_global[1]))
         sys.stdout.write("{:<7} {:<13}{:>10.2e} {:<13}{:>10.2e} {:<14}{:>10.2e}\n".format("Min:", min_us[0], min_us[1], \
             min_europe[0], min_europe[1], min_global[0], min_global[1]))
    elif args[0] == "Process Energy":
        energy = args[1]
        sys.stdout.write("-"*80+ "\n" + "-"*80 + "\n")
        sys.stdout.write("{:<13} {:51} {:>10.2e} {:>3}\n".format("Process used:", "", energy, "kWh"))

    else:
        sys.stdout.write(args[0])


""" MISC UTILS """

def get_data(file):
    file = os.path.join(DIR_PATH, file)
    with open(file) as f:
            data = json.load(f)
    return data

def valid_cpu():
    return os.path.exists(BASE) and bool(os.listdir(BASE))

def valid_gpu():
    """ Checks that there is a valid Nvidia GPU """


    try:
        bash_command = "nvidia-smi > /dev/null 2>&1" #we must pipe to ignore error message
        output = subprocess.check_call(['bash','-c', bash_command])
        if type(output) == float:
            return True
        else:
            return False

    except:
        return False
